source/Plugins.cpp

```diff
@@ -38,14 +64,135 @@ namespace {
+	{
+		int retVal;
+		const void *buff;
+		size_t size;
+		int64_t offset;
+
+		while(true)
+		{
+			retVal = archive_read_data_block(ar, &buff, &size, &offset);
+			if(retVal == ARCHIVE_EOF)
+				return;
+			if(retVal != ARCHIVE_OK)
+				return;
+			if(archive_write_data_block(aw, buff, size, offset) != ARCHIVE_OK)
+				return;
+		}
+	}
+
+	bool ExtractZIP(std::string &filename, const std::string &destination, const std::string &expectedName)
+	{
+		int flags = ARCHIVE_EXTRACT_TIME;
+		flags |= ARCHIVE_EXTRACT_PERM;
+		flags |= ARCHIVE_EXTRACT_ACL;
+		flags |= ARCHIVE_EXTRACT_FFLAGS;
+
+		// Create the handles for reading/writing
+		archive *read = archive_read_new();
+		archive *ext = archive_write_disk_new();
+		archive_write_disk_set_options(ext, flags);
+		archive_read_support_format_all(read);
+		if(!filename.empty() && filename.front() == '-')
+			filename.clear();
+		if(archive_read_open_filename(read, filename.c_str(), 10240))
+			return false;
+
+		// Check if this plugin has the right head folder name
+		archive_entry *entry;
+		archive_read_next_header(read, &entry);
+		string firstEntry = archive_entry_pathname(entry);
+		firstEntry = firstEntry.substr(0, firstEntry.find("/")) + "/";
+		bool fitsExpected = firstEntry == (expectedName);
+		archive_read_data_skip(read);
+
+		// Check if this plugin has a head folder, if not create one in the destination
+		archive_read_next_header(read, &entry);
+		string secondEntry = archive_entry_pathname(entry);
+		bool hasHeadFolder = secondEntry.find(firstEntry) != std::string::npos;
+		if(!hasHeadFolder)
+#if defined(_WIN32)
+			_wmkdir(Utf8::ToUTF16(destination + expectedName).c_str());
+#else
+			mkdir((destination + expectedName).c_str(), 0777);
+#endif
+
+		// Close the archive so we can start again from the beginning
+		archive_read_close(read);
+		archive_read_free(read);
+
+		// Read another time, this time for writing.
+		read = archive_read_new();
+		archive_read_support_format_all(read);
+		archive_read_open_filename(read, filename.c_str(), 10240);
+
+		int size = 0;
+		while (true)
+		{
+			int retVal = archive_read_next_header(read, &entry);
+			if(retVal == ARCHIVE_EOF)
+				break;
+			if(retVal != ARCHIVE_OK)
+				return false;
+
+			size += archive_entry_size(entry);
+			// The extracted size may be 4 times the maximum download size.
+			if(size > MAX_DOWNLOAD_SIZE * 4)
+				return false;
+
+			// Adjust root folder name if neccessary.
+			if(!fitsExpected && hasHeadFolder)
+			{
+				string thisEntryName = archive_entry_pathname(entry);
+				size_t start_pos = thisEntryName.find(firstEntry);
+				if(start_pos != std::string::npos)
+					thisEntryName.replace(start_pos, firstEntry.length(), expectedName);
+
+				archive_entry_set_pathname(entry, thisEntryName.c_str());
+			}
+
+			// Add root folder to path if neccessary.
+			string dest_file = (destination + (hasHeadFolder ? "" : expectedName)) + archive_entry_pathname(entry);
```

Comment:
If I'm right, this is also vulnerable to zip slip attacks.

Reply:
explain please

Reply:
Essentially, it allows placing files outside of the directory where you begin unzipping a file. It's a path traversal attack, basically the same thing I described above.

Reply:
I added a check for ".."

Reply:
That should be good enough here, unless there is some magic on some OSes with drive letters or environment variable expansion stuff...

When using the plugin name, make sure to check for single dots as well, as we don't want to use unzip into the 'plugins' directory.

Reply:
so check if the plugin name equals a single dot?

Reply:
i should probably also check for a / or a backslash